### ISSUE TITLE
VB & C# cannot consume method with modopt

### ISSUE THREAD

**Issue by DzonnyDZ** [dotnet/roslyn#4163]
I have a C++ library that defines a member like
```C++/CLI
property Nullable<DWORD> Type{Nullable<DWORD> get();}
```
which is translated to CLI as
```CIL
.property instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> Type
{
    .get instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> Dzonny.XmlKeyboard.Interop.KbdTables::get_Type()
}
```

VB complains that it cannot consume it because "Property 'Type' is of an unsupported type."

This used to work before Roslyn, and it should work according to CTS specification consumer must read and correctly propagate all modopts, even if it does not understand them.

C# has the same issue, but with more misleading error message:  "Property, indexer, or event 'KbdTables.Type' is not supported by the language; try directly calling accessor method 'KbdTables.get_Type()'." But when I try to call get_Type() directly I get: "'KbdTables.get_Type()' is not supported by the language."

**Comment by gafter** [comment]
Here is a draft of a unit test for the C# version of this:

```cs
        /// <summary>
        /// Test invoking properties involving custom modifiers
        /// </summary>
        [Fact]
        public void InvokePropertyWithCustomModifier()
        {
            var il = @"
.assembly extern mscorlib { .ver 4:0:0:0 .publickeytoken = (B7 7A 5C 56 19 34 E0 89) }
.assembly extern System.Core {}

.assembly '<<GeneratedFileName>>'
{
}

.class public auto ansi beforefieldinit CustomProps
       extends [mscorlib]System.Object
{
  .method public hidebysig newslot specialname virtual instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)>
          get_Prop() cil managed
  {
    // Code size       3 (0x3)
    .maxstack  8
    IL_0000:  nop
    IL_0001:  ldnull
    IL_0002:  throw
  } // end of method CustomProps::get_Prop

  .property instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> Prop()
  {
      .get instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> CustomProps::get_Prop()
  }
}
";

            var source = @"
class C
{
    void M(CustomProps p)
    {
        var x = p.Prop;
    }
}
";
            var ilRef = CompileIL(il, appendDefaultHeader: false);
            var comp = CreateCompilation(source, new[] { MscorlibRef, SystemCoreRef, ilRef });
            comp.VerifyDiagnostics();

            var global = comp.GlobalNamespace;
            var getter = global.GetMember<NamedTypeSymbol>("CustomProps").GetMethod("get_Prop");
            Assert.Equal("TODO", getter.ToTestDisplayString());
        }
```

**Comment by gafter** [comment]
While investigating this, I discovered that the root cause is that this kind of code is not expressible in Roslyn's internal data structures, so it just produces "error" symbols from the IL.

Fixing this and related issues requires adding the following API elements (and implementing it throughout both compilers):

```cs
/// <summary>
/// An array corresponding to the type arguments, each element of which contains the
// custom modifiers associated with the corresponding type argument,
/// or an empty array if there are none for that argument.
/// </summary>
ImmutableArray<ImmutableArray<CustomModifier>> INamedTypeSymbol.TypeArgumentsCustomModifiers { get; }

/// <summary>
/// Returns a constructed type given its type arguments and custom modifiers for its type arguments.
/// </summary>
/// <param name="typeArguments">The immediate type arguments to be replaced for type
/// parameters in the type.</param>
INamedTypeSymbol Construct(params ITypeSymbol[] typeArguments, ImmutableArray<ImmutableArray<CustomModifier>> typeArgumentsCustomModifiers);
```

There is a related issue for events, which would require adding

```cs
/// <summary>
/// Custom modifiers associated with the event's type, or an empty array if there are none.
/// </summary>
ImmutableArray<CustomModifier> IEventSymbol.CustomModifiers { get; }
```

I'm ignoring that for now.

/cc @AnthonyDGreen @jaredpar @VSadov

**Comment by AlekseyTs** [comment]
According to ECMA-335 (Section II.23.2.12 Type), type arguments of a generic instantiation are encoded as ```Type```.
```
Type ::=
BOOLEAN | CHAR | I1 | U1 | I2 | U2 | I4 | U4 | I8 | U8 | R4 | R8 | I | U
| ARRAY Type ArrayShape (general array, see §II.23.2.13)
| CLASS TypeDefOrRefOrSpecEncoded
| FNPTR MethodDefSig
| FNPTR MethodRefSig
| GENERICINST (CLASS | VALUETYPE) TypeDefOrRefOrSpecEncoded GenArgCount Type*
| MVAR number
| OBJECT
| PTR CustomMod* Type
| PTR CustomMod* VOID
| STRING
| SZARRAY CustomMod* Type (single dimensional, zero-based array i.e.,vector)
| VALUETYPE TypeDefOrRefOrSpecEncoded
| VAR number
```

According to Section II.23.2.14 TypeSpec
The signature in the Blob heap indexed by a TypeSpec token has the following format –
```
TypeSpecBlob ::=
PTR CustomMod* VOID
| PTR CustomMod* Type
| FNPTR MethodDefSig
| FNPTR MethodRefSig
| ARRAY Type ArrayShape
| SZARRAY CustomMod* Type
| GENERICINST (CLASS | VALUETYPE) TypeDefOrRefOrSpecEncoded GenArgCount Type Type*
```

In this context, custom modifiers are valid only as part of PTR and SZARRAY, uint32 is neither.

**Comment by AlekseyTs** [comment]
Assigned to @gafter per offline conversation.

**Comment by DzonnyDZ** [comment]
Uh, I must admit that I'm somewhat confused by the CIL specification. Yes, in II.23.2.14 it says what AlekseyTs posted. I was hoping that TypeDefOrRefOrSpecEncopded would contain CustomMod, but no :-O. On the other hand in II.7.1 the spec says (among others) 
```
Type ::= 
| Type modopt ‘(’ TypeReference ‘)’ 
| Type modreq ‘(’ TypeReference ‘)’ 
```
I'm not sure how these two parts of the spec relate.
Another thing to consider is that the given C++/CLI program exists and runs. How comes? Flaw in the spec? C++/CLI not conforming to the spec? But how does .NET than understand the DLL? Microsoft-specific (woun't the code run on e.g. Mono?)?

**Comment by AlekseyTs** [comment]
Section II.7.1 is about IL grammar. Custom modifiers exist purely for compilers, CLR doesn’t need to understand them as long as two signatures match exactly. That is probably the reason why existing implementations of runtime do not enforce specification, which is unfortunate.

**Comment by gafter** [comment]
I will document this as a known incompatibility.

**Comment by gafter** [comment]
@DzonnyDZ Clearly the section you quote (II.7.1) in incorrect, as modifiers precede (rather than follow) the modified type.

**Comment by davkean** [comment]
In my opinion, the spec is wrong with its representation of this (which is why very little metadata readers represent the type system like in the spec[1]). Doing so, would make it impossible to represent a number of valid C++ constructs, such as the following:

``` C++
	void Foo(Nullable<int> value)
	{
	}

	void Foo(Nullable<long> value)
	{
	}
```
//cc @nguerrera @CarolEidt for their opinions.

[1] CCI, for example, treats modifiers just like other constructed types (such as arrays and pointers) where a modifier is a "Type" that has a single element of "Type".

**Comment by nguerrera** [comment]
I share the same opinion as David here. Whether intentional or not, ECMA-335 excludes cases that are useful (such as David's example) and work fine in practice. I suggest that we amend the spec to allow those cases.

**Comment by CarolEidt** [comment]
It seems reasonable to modify the spec to allow these cases to be supported. It would be great if someone (@davkean?) could propose a specific change, presumably to II.23.2.12 & II.23.2.14. The task group is currently on hiatus, but it would be good to have this formulated as a specific proposal.

**Comment by davkean** [comment]
@nguerrera had an ideal on simplifying this entire section (as well as Param, FieldSig, PropertySig, etc) to basically treat modifiers (as well as byref) as simply valid "Types" (because that's exactly how you decode them) and then document the invalid states (ie byref must be last).

I'll talk to Nick offline, and between the two of us, should be able to come up a proposal.

**Comment by AlekseyTs** [comment]
@gafter Please verify.
